File: Arduino/firmvare v2/src/DipLink.cpp
#include "DipLink.h"

int DipLink::getNumLen(int number)
{
  int buf = number;

  int count = (buf == 0) ? 1 : 0;
  while (buf != 0)
  {
      count++;
      buf /= 10;
  }

  return count;
}

String DipLink::zeroAdd(String number)
{
  int count = getNumLen(number.length());

  String msgLenFinal;

  switch (count)
  {
    case 1:
      msgLenFinal = msgLenFinal + "00" + number.length();
      break;
    case 2:
      msgLenFinal = msgLenFinal + '0' + number.length();
      break;
    default:
      break;
  }

  return msgLenFinal;
}

DipLink::DipLink()
{
  connectionEnabled = false;
}

DipLink::~DipLink()
{
  if(connectionEnabled)
  {
    String message = "55013001ENDCONNECTION";
    message += message.length();
    Serial.print(message);
    Serial.end();
  }
}

int DipLink::enableConnection()
{
  if(!connectionEnabled)
  {
    Serial.begin(9600);

    while (true)
    {
      while (Serial.available() <= 0)
      {
        Serial.println(createHeartBeatMessage());
        delay(2000);
      }

      String message = Serial.readString();
      if(readMessage(message).equals("000_BEAT"))
      {
        connectionEnabled = true;
        return 0;// connection enabled
      }
    }
  }

  return 2;//major error
}

int DipLink::checkConnection()
{
  if (connectionEnabled)
  {
    int count;

    while(true)
    {
      count = 0;

      while (Serial.available() <= 0 and count <= 3)
      {
          Serial.print(createHeartBeatMessage());
          count++;
      }

      if(count > 3)
      {
        connectionEnabled = false;
        return 1; //connection break
      }
      else
      {
        String message = Serial.readString();
        if(readMessage(message).equals("000_BEAT"))
        {
          return 0; //connection OK
        }
      }
    }
  }
  else
  {
    connectionEnabled = false;
    return 1; //connection break
  }

  return 2; //major error
}

int DipLink::reconnect()
{
  Serial.end();
  if (enableConnection() != 0)
  {
    return 0;
  }

  return 2; //major error
}

String DipLink::readMessage(String message)
{
  String line = message;
  String resMessage;
  String buf;

  line.trim();

  //get package start info
  for (int i = 0; i < 2; i++)
  {
    buf += line[i];
  }

  if (!buf.equals("55"))
  {
    return "NOT_PACKAGE";
  }

  buf = "";
  line = line.substring(2); //delete package start info

  //get msgLen info
  for (int i = 0; i < 3; i++)
  {
    buf += line[i];
  }

  if (buf[0] == '0')
  {
    if (buf[1] == '0')
    {
      buf.substring(2);
    }
    else
    {
      buf.substring(1);
    }
  }

  int msgLen;
  if ((msgLen = buf.toInt()) == 0)
  {
    return "MSGLEN_ERROR";
  }

  buf = "";
  line = line.substring(3); //delete msgLen info

  //parsing msgID
  for (int i = 0; i < 3; i++)
  {
    buf += line[i];
  }

  resMessage = resMessage + buf + '_';

  buf = "";
  line = line.substring(3); //delete msgID info

  //parsing msg info
  for (int i = 0; i < msgLen; i++)
  {
    buf += line[i];
  }

  resMessage = resMessage + buf;

  buf = "";
  line = line.substring(msgLen); //delete msg info

  //parsing packageLen info
  for (unsigned int i = 0; i < line.length() + 1; i++)
  {
    buf += line[i];
  }

  unsigned int packageLen;
  if ((packageLen = buf.toInt()) == 0)
  {
    return "PACKAGELEN_ERROR";
  }

  //chek for package integrity
  if (packageLen != message.length())
  {
    return "PACKAGE_BROKEN";
  }

  return resMessage;
}

String DipLink::writeMessage(String message, String msgID)
{
  String package = "55";

  String msgLen = zeroAdd(message);

  package = package + msgLen + msgID + message;
  package += (package.length() + getNumLen(package.length()));
  package.toUpperCase();

  return package;
}

String DipLink::createHeartBeatMessage()
{
  return writeMessage("BEAT", "000");
}

int DipLink::sendMessage(String message)
{
  if(connectionEnabled)
  {
    Serial.println(message);
    return 0;
  }
  return 1;//no connection
}

String DipLink::getMessage()
{
  if (connectionEnabled)
  {
    String message = Serial.readString();
    Serial.println(message);
    return message;
  }
  return "";
}

int DipLink::getCommandID(String command)
{
  String commandID;
  int i = 0;

  while ((command[i] == 0) and i < 3)
  {
    i++;
  }

  while(command[i] != '_')
  {
    commandID += command[i];
    i++;
  }

  if(commandID.toInt() != 0)
  {
    return commandID.toInt();
  }
  else
  {
    return 1;//not numberic value
  }
}

String getCommandMsg(String command)
{
  String commandName;

  for(unsigned int i = 4; i < command.length(); i++)
  {
    commandName += command[i];
  }

  return commandName;
}

File: Android/ArduinoConnectV2/app/src/main/java/com/diploma/mattue/arduinoconnectv2/DipLink.java
package com.diploma.mattue.arduinoconnectv2;

import android.content.BroadcastReceiver;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.ServiceConnection;
import android.os.Bundle;
import android.os.Handler;
import android.os.IBinder;
import android.os.Message;
import android.widget.Toast;

import java.lang.ref.WeakReference;
import java.util.Set;

/**
 * Created by Mattue on 05.06.2017.
 */

public class DipLink
{
    public DipLink()
    {
        connectionEnabled = false;
    }

    public int enableConnection()
    {
        if (!connectionEnabled)
        {
            //enable connection here
        }
        else
        {
            return 0; //already connected
        }

        return 2; //major error
    }

    public int chekConnection()
    {
        if (connectionEnabled)
        {
            //check connection here
        }
        else
        {
            return 1; //connection break
        }

        return 2; //major error
    }

    public int reconnect()
    {
        //close connection here

        return enableConnection();
    }

    public String readMessage(String message)
    {
        int msgLen = 0;
        String line = message;
        String resMessage = "";
        line.trim();

        String buf = "";

        //get package start info
        buf = line.substring(0,1);
        if(!buf.equalsIgnoreCase("55"))
        {
            return "NOT_PAKAGE";
        }

        //get msgLen info
        buf = line.substring(2,4);
        try
        {
            msgLen = Integer.parseInt(buf);
        }
        catch(NumberFormatException e)
        {
            return "MSG_LEN_FORMAT_ERROR";
        }

        //get msgID info
        buf = line.substring(5, 7);
        try
        {
            Integer.parseInt(buf);
        }
        catch(NumberFormatException e)
        {
            return "MSG_ID_FORMAT_ERROR";
        }
        resMessage = resMessage + buf + '_';

        //get msg info
        buf = line.substring(8, msgLen);
        resMessage += buf;

        //get packageLen
        buf = line.substring(msgLen + 1);
        try
        {
            Integer.parseInt(buf);
        }
        catch(NumberFormatException e)
        {
            return "PACKAGE_LEN_ERROR";
        }

        //check for package integrity
        if (Integer.parseInt(buf) != message.length())
        {
            return "PACKAGE_BROKEN";
        }

        return resMessage;
    }

    public String writeMessage(String msgID, String message)
    {
        String messagePackage = "55";

        messagePackage = messagePackage + zeroAdd(message) + msgID + message;
        messagePackage += (messagePackage.length() + getNumLen(messagePackage.length()));
        messagePackage.toUpperCase();

        return messagePackage;
    }

    public String createHeartBeatMessage()
    {
        return writeMessage("000", "BEAT");
    }

    private int getNumLen(int length)
    {
        int buf = length;

        int count = (buf == 0) ? 1 : 0;
        while (buf != 0)
        {
            count ++;
            buf /= 10;
        }

        return count;
    }

    public int getCommandID(String commamd)
    {
        String commandID;

        commandID = commamd.substring(0, 2);
        try
        {
            return Integer.parseInt(commandID);
        }
        catch (NumberFormatException e)
        {
            return 1; //not numeric value
        }
    }

    public String getCommandMsg(String command)
    {
        return command.substring(4);
    }

    //определяет длину текста "BEAT" и выдает длину этого текста в нулями в переди
    private String zeroAdd(String number)
    {
        int count = getNumLen(number.length());
        String resString = "";

        switch (count)
        {
            case 1:
                resString = resString + "00" + number.length();
                break;
            case 2:
                resString = resString + '0' + number.length();
                break;
            default:
                break;
        }

        return resString;
    }

    private boolean connectionEnabled;
}
